### ISSUE TITLE
instanceof redundant for class, but not for interface

### ISSUE THREAD

**Issue by donquixote** [vimeo/psalm#7341]
psalm is inconsistent in when it says "RedundantCondition" on instanceof checks.

https://psalm.dev/r/377cb86ce7

```php
interface I {}
class C {}

/**
 * @param I $o
 */
function fI(I $o): void {
    // Not redundant?
    if ($o instanceof I) {
        print 'yes';
    }
}

/**
 * @param C $o
 */
function fC(C $o): void {
    // ERROR: RedundantCondition - 20:9 - Type C for $o is always C
    if ($o instanceof C) {
        print 'yes';
    }
}
```

**Comment by orklah** [comment]
The difference is here: https://github.com/vimeo/psalm/blob/a53b5215b88cd57e53bb2acb2038dd3230c96f8c/src/Psalm/Internal/Type/AssertionReconciler.php#L504

In your snippet, it doesn't matter that much, but the RedundantCondition is a false positive because of this: https://psalm.dev/r/3bd23090dd

**Comment by orklah** [comment]
I actually worked on a very similar issue very close to this code: https://github.com/vimeo/psalm/pull/6739

In fact, when dropping the `&& !$new_type_has_interface` I linked above, it's my test that fails...

Not sure why there are two conditions
